src/shogun/regression/LinearRidgeRegression.cpp

```diff
@@ -57,32 +83,26 @@ bool CLinearRidgeRegression::train_machine(CFeatures* data)
     	"Expected Real Features (%d) but got (%d)\n",
     	F_DREAL, data->get_feature_type());
 
-    CDenseFeatures<float64_t>* feats=(CDenseFeatures<float64_t>*) data;
-    int32_t num_feat=feats->get_num_features();
-
-    SGMatrix<float64_t> kernel_matrix(num_feat,num_feat);
-    SGMatrix<float64_t> feats_matrix(feats->get_feature_matrix());
-    SGVector<float64_t> y(num_feat);
-    SGVector<float64_t> tau_vector(num_feat);
-
-    tau_vector.zero();
-    tau_vector.add(m_tau);
-
-	linalg::matrix_prod(feats_matrix, feats_matrix, kernel_matrix, false, true);
-	linalg::add_diag(kernel_matrix, tau_vector);
-
-	auto labels = regression_labels(m_labels);
-	auto lab = regression_labels(m_labels)->get_labels();
-	linalg::matrix_prod(feats_matrix, lab, y);
-
-	auto decomposition = linalg::cholesky_factor(kernel_matrix);
-	y = linalg::cholesky_solver(decomposition, y);
-	set_w(y);
-
-	auto x_bar = linalg::colwise_sum(feats_matrix);
-	linalg::scale(x_bar, x_bar, 1.0 / ((float64_t)x_bar.size()));
-	auto intercept = linalg::mean(lab) - linalg::dot(lab, x_bar);
-	set_bias(intercept);
+	auto feats = data->as<CDenseFeatures<float64_t>>();
+	auto y = regression_labels(m_labels)->get_labels();
+	auto N = feats->get_num_vectors();
+	auto D = feats->get_num_features();
+
+	SGVector<float64_t> w;
+	if (N >= D)
+		w = solve_using_covariance<float64_t>(feats, y, m_tau);
+	else
+		w = solve_using_gram<float64_t>(feats, y, m_tau);
+
+	set_w(w);
+
+	if (m_use_bias)
+	{
+		auto x_bar = feats->sum();
+		linalg::scale(x_bar, x_bar, ((float64_t)1.0) / N);
+		auto intercept = linalg::mean(y) - linalg::dot(w, x_bar);
```

Comment:
For consistency in the computations of features and labels, is it possible to do some `linalg::mean` of features?